### ISSUE TITLE
pickling quaternions fails 

### ISSUE THREAD

**Issue by ulijh** [moble/quaternion#64]
```
import quaternion
import pickle

a = quaternion.from_rotation_vector([0, 0, np.pi/3])

pickle.dumps(a)
```
produces:
```
---------------------------------------------------------------------------
PicklingError                             Traceback (most recent call last)
<ipython-input-47-143f409b05ff> in <module>()
      5 a = quaternion.from_rotation_vector([0, 0, np.pi/3])
      6 
----> 7 pickle.dumps(a)

PicklingError: Can't pickle <class 'quaternion'>: attribute lookup quaternion on builtins failed
```
I would like to use quaternions inside an application using dask, [http://dask.pydata.org/en/latest/] and dask distributed which seems to rely on cloudpickle to distribute work to workers on a cluster.

How hard would it be to get this working?

Thanks!

**Comment by moble** [comment]
You should *really* search issues before you open a new issue.  See #14.  However, also note the workaround I gave [here](https://github.com/moble/quaternion/issues/41#issuecomment-287515017).

**Comment by ulijh** [comment]
You got me again, I missed #14, but I read the workaround you mentioned. However this does not fit in my workflow. It's not about saving data to disk. Thanks any ways.

**Comment by moble** [comment]
The key point about the workaround in #14 is not the saving to disk; it's the conversion of an array of quaternions into and then from an array of doubles, which can be pickled.  Again, I don't know anything about your workflow, but it seems like any problem that could be solved by pickling quaternion arrays could just incorporate these two extra steps around the pickle/unpickle steps.

**Comment by ulijh** [comment]
My workflow is as follows: I have an expensive coordinate transformation for many points which then are interpolated onto a regular grid. The coordinate transformations currently involve your quaternions because they are fast and accurate and integrate nicely. Now I want to do this in parallel using dask.distributed [https://github.com/dask/distributed](url) which pickles some stuff (I don't know the details) and I just can't go inside and ad some: If is quaternion, convert to numpy-array. Thats the background for this question.

**Comment by moble** [comment]
So you're giving `dask` some functions that either take quaternions as inputs or return quaternions as outputs.  I haven't used `dask` in a while, but as I recall, it just pickles things to move data around, and then calls your functions but doesn't care what happens inside those functions.  So couldn't you just rewrite your functions to take numpy arrays as inputs, convert them to quaternions, do your work, and then convert them again to arrays on output?  The conversion step really is very fast because (almost) no memory is being allocated or copied, it's just a different numpy view on the same data.

**Comment by ulijh** [comment]
Well, yeah, I should give this a try. Thanks! ;-)

**Comment by belm0** [comment]
another workaround:
```python
>>> import builtins, quaternion, pickle
>>> builtins.quaternion = quaternion.quaternion
>>> b = pickle.dumps(quaternion.one)
>>> pickle.loads(b)
quaternion(1, 0, 0, 0)
```

**Comment by moble** [comment]
Interesting.  I didn't know that trick.  But I'll just point out that if you want to load the pickle from another process (e.g., after storing to file and starting another python session), you have to run the first two lines again.  So, for posterity, the process is this:
```python
>>> import builtins, quaternion, pickle
>>> builtins.quaternion = quaternion.quaternion
>>> with open('pickled_quaternion.pickle', 'wb') as f:
...     pickle.dump(quaternion.one, f)
```
[start a new python session]
```python
>>> import builtins, quaternion, pickle
>>> builtins.quaternion = quaternion.quaternion
>>> with open('pickled_quaternion.pickle', 'rb') as f:
...     q = pickle.load(f)
>>> q                                                                                                                                                                               
quaternion(1, 0, 0, 0)
```

**Comment by belm0** [comment]
I wonder if the underlying problem can be addressed within numpy_quaternion.c

Some hints at https://stackoverflow.com/questions/4677012/python-cant-pickle-type-x-attribute-lookup-failed

**Comment by belm0** [comment]
I believe the problem is in the configuration of `tp_name`.

https://github.com/moble/quaternion/blob/a0fdc4ff954514de12336a26532b594387146299/numpy_quaternion.c#L806-L813

> **For types that are accessible as module globals, the string should be the full module name, followed by a dot, followed by the type name**; for built-in types, it should be just the type name.

https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_name
